packages/components/src/font-size-picker/index.tsx

```diff
@@ -201,141 +192,145 @@ const UnforwardedFontSizePicker = (
-								{ withReset && (
-									<FlexItem>
-										<ResetButton
-											disabled={ value === undefined }
-											onClick={ () => {
-												onChange?.( undefined );
-											} }
-											isSmall
-											variant="secondary"
-											size={ size }
-										>
-											{ __( 'Reset' ) }
-										</ResetButton>
-									</FlexItem>
-								) }
-							</Flex>
-						) }
-				</Controls>
-				{ withSlider && (
-					<RangeControl
+				{ ! shouldUseSelectControl && ! showCustomValueControl && (
+					<ToggleGroupControl
 						__nextHasNoMarginBottom={ __nextHasNoMarginBottom }
-						className="components-font-size-picker__custom-input"
-						label={ __( 'Custom Size' ) }
-						value={
-							isPixelValue && noUnitsValue
-								? Number( noUnitsValue )
-								: undefined
-						}
-						initialPosition={ fallbackFontSize }
+						label={ __( 'Font size' ) }
+						hideLabelFromVision
+						value={ value }
 						onChange={ ( newValue ) => {
-							onChange?.( hasUnits ? newValue + 'px' : newValue );
+							if ( newValue === undefined ) {
+								onChange?.( undefined );
+							} else {
+								onChange?.(
+									hasUnits ? newValue : Number( newValue )
+								);
+							}
 						} }
-						min={ 12 }
-						max={ 100 }
-					/>
+						isBlock
+						size={ size }
+					>
+						{ ( options as FontSizeToggleGroupOption[] ).map(
+							( option ) => (
+								<ToggleGroupControlOption
+									key={ option.key }
+									value={ option.value }
+									label={ option.label }
+									aria-label={ option.name }
+									showTooltip={ true }
+								/>
+							)
+						) }
+					</ToggleGroupControl>
+				) }
+				{ ! disableCustomFontSizes && showCustomValueControl && (
+					<Flex className="components-font-size-picker__custom-size-control">
+						<FlexItem isBlock>
+							<UnitControl
+								label={ __( 'Custom' ) }
+								labelPosition="top"
+								hideLabelFromVision
+								value={ value }
+								onChange={ ( newValue ) => {
+									if ( newValue === undefined ) {
+										onChange?.( undefined );
+									} else {
+										onChange?.(
+											hasUnits
+												? newValue
+												: parseFloat( newValue )
+										);
+									}
+								} }
+								size={ size }
+								units={ hasUnits ? units : [] }
+							/>
+						</FlexItem>
+						{ withSlider && (
+							<FlexItem isBlock>
+								<RangeControl
+									__nextHasNoMarginBottom={
+										__nextHasNoMarginBottom
+									}
+									className="components-font-size-picker__custom-input"
+									label={ __( 'Custom Size' ) }
+									hideLabelFromVision
+									value={ valueNumber }
+									initialPosition={ fallbackFontSize }
+									withInputField={ false }
+									onChange={ ( newValue ) => {
+										if ( newValue === undefined ) {
+											onChange?.( undefined );
+										} else if ( hasUnits ) {
+											onChange?.(
+												newValue + ( valueUnit ?? 'px' )
```

Comment:
Just double-checking, it looks like `valueUnit` will always be a string or undefined, right? That means we can safely use `??` because it'll never accidentally concatenate `false` or `null`?

Reply:
That's right. You can hover over `valueUnit` and TypeScript will tell you that it's `string | undefined`.